lib/view/sheet/decleration.py

```diff
@@ -380,12 +381,35 @@ def max(initializer=None, converter=None):
         )
 
 
-class BaseProjector(object):
-    field_type = None  # required override
-    source = None  # optional override
-    keys = None  # optional override
+class Projector(object):
+    def __init__(self, field_type: str):
+        """
+        Arguments:
+        source -- Name of the source to project from.
+        """
+        self.field_type = field_type
+
+    def __call__(self, sheet, sources):
+        try:
+            result = self.do_project(sheet, sources)
+        except (NoEntryException, ErrorEntryException):
+            raise
+        except Exception as e:
+            # XXX - A debug log may be useful.
+            # print 'debug - ', e, self.source, self.source
+            raise ErrorEntryException("unexpected error occurred: {}".format(e))
+
+        return result
+
+    def do_project(self, sheet, sources):
+        raise NotImplementedError("override do_project")
 
-    def __init__(self, source, *keys, **kwargs):
+    def get_sources(self):
+        raise NotImplementedError("override get_sources")
```

Comment:
This is the main thing that started the refactor. I needed a single method to get sources from different types.